File: solutions/prospective/algorithms/patterns/c/dfs_recursive_grid.c
/*
DFS on Grid (Recursive) Template

Standard flood fill or island counting pattern.
- Check bounds.
- Check visited/condition.
- Recurse neighbors.
*/
// === MEMO START ===
void dfs(char **grid, int rows, int cols, int r, int c) {
    if (r < 0 || c < 0 || r >= rows || c >= cols || grid[r][c] == '0') {
        return;
    }

    grid[r][c] = '0'; // Mark visited

    dfs(grid, rows, cols, r + 1, c);
    dfs(grid, rows, cols, r - 1, c);
    dfs(grid, rows, cols, r, c + 1);
    dfs(grid, rows, cols, r, c - 1);
}

int num_islands(char **grid, int rows, int cols) {
    if (!grid) return 0;
    int count = 0;

    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == '1') {
                count++;
                dfs(grid, rows, cols, r, c);
            }
        }
    }

    return count;
}

File: solutions/prospective/algorithms/patterns/lua/dfs_recursive_grid.lua
--[[
DFS on Grid (Recursive) Template

Standard flood fill or island counting pattern.
- Check bounds.
- Check visited/condition.
- Recurse neighbors.
--]]
-- === MEMO START ===
function num_islands(grid)
    if #grid == 0 then return 0 end
    local rows = #grid
    local cols = #grid[1]
    local count = 0
    
    local function dfs(r, c)
        if r < 1 or c < 1 or r > rows or c > cols or grid[r][c] == "0" then
            return
        end
        
        grid[r][c] = "0" -- Mark visited
        
        dfs(r + 1, c)
        dfs(r - 1, c)
        dfs(r, c + 1)
        dfs(r, c - 1)
    end

    for r = 1, rows do
        for c = 1, cols do
            if grid[r][c] == "1" then
                count = count + 1
                dfs(r, c)
            end
        end
    end
    
    return count
end

